app/src/main/java/org/p2p/wallet/home/ui/crypto/MyCryptoPresenter.kt

```diff
@@ -0,0 +1,195 @@
+    private val connectionManager: ConnectionManager,
+    private val transactionManager: TransactionManager,
+    private val tokenServiceCoordinator: TokenServiceCoordinator,
+    private val analytics: HomeAnalytics,
+) : BasePresenter<MyCryptoContract.View>(), MyCryptoContract.Presenter {
+
+    private var visibilityState: VisibilityState = VisibilityState.Hidden
+    private var cryptoTokensSubscription: Job? = null
+
+    override fun attach(view: MyCryptoContract.View) {
+        super.attach(view)
+        initializeActionButtons()
+        observeCryptoTokens()
+    }
+
+    override fun refreshTokens() {
+        launchInternetAware(connectionManager) {
+            try {
+                tokenServiceCoordinator.refresh()
+            } catch (cancelled: CancellationException) {
+                Timber.i("Loading tokens job cancelled")
+            } catch (error: Throwable) {
+                Timber.e(error, "Error refreshing user tokens")
+                view?.showErrorMessage(error)
+            }
+        }
+    }
+
+    private fun observeCryptoTokens() {
+        cryptoTokensSubscription?.cancel()
+        cryptoTokensSubscription = launch {
+            tokenServiceCoordinator.observeUserTokens()
+                .collect { handleTokenState(it) }
+        }
+    }
+
+    private fun handleTokenState(newState: UserTokensState) {
+        view?.showRefreshing(isRefreshing = newState.isLoading())
+
+        when (newState) {
+            is UserTokensState.Idle -> Unit
+            is UserTokensState.Loading -> Unit
+            is UserTokensState.Refreshing -> Unit
+            is UserTokensState.Error -> {
+                view?.showErrorMessage(newState.cause)
+            }
+            is UserTokensState.Empty -> {
+                view?.showEmptyState(isEmpty = true)
+                handleEmptyAccount()
+            }
+            is UserTokensState.Loaded -> {
+                view?.showEmptyState(isEmpty = false)
+                initializeActionButtons()
+                showTokensAndBalance(
+                    solTokens = newState.solTokens,
+                    ethTokens = newState.ethTokens
+                )
+            }
+        }
+    }
+
+    private fun showTokensAndBalance(solTokens: List<Token.Active>, ethTokens: List<Token.Eth>) {
+        val balance = getUserBalance(solTokens)
+        view?.showBalance(homeMapper.mapBalance(balance))
+        logBalance(balance)
+
+        val areZerosHidden = homeInteractor.areZerosHidden()
+        val mappedItems: List<AnyCellItem> = homeMapper.mapToCellItems(
+            tokens = solTokens,
+            ethereumTokens = ethTokens,
+            visibilityState = visibilityState,
+            isZerosHidden = areZerosHidden,
+        )
+        view?.showItems(mappedItems)
+    }
+
+    private fun getUserBalance(tokens: List<Token.Active>): BigDecimal {
+        if (tokens.none { it.totalInUsd != null }) return BigDecimal.ZERO
+
+        return tokens
```

Comment:
UserRepository already has this logic

Reply:
yes but we will need to filter this tokens on USDC and other Wallet tokens in future